IFC-761 Use flags to handle branch permissions

gmazoyer opened the pull request.

gmazoyer committed:
Use flags to represent per branch permission

gmazoyer committed:
Remove `branch` from object permission objects

gmazoyer committed:
Fix circular dependency

gmazoyer committed:
Fix error message to match

gmazoyer committed:
Adapt local permission backend to handle flags

gmazoyer committed:
Fix some programming errors

gmazoyer committed:
Fix object permission protocol

ogenstad commented on backend/infrahub/core/constants/__init__.py:
```diff
@@ -69,9 +69,11 @@ class PermissionAction(InfrahubStringEnum):
+    DENY = 1
+    ALLOWED_DEFAULT = 2
+    ALLOWED_OTHER = 4
+    ALLOWED_ALL = 6
```
Sorry I realize this is probably my fault as I typed this somewhere. but perhaps we should change it to:

```python
class PermissionDecision(InfrahubNumberEnum):
    DENY = 1
    ALLOW_DEFAULT = 2
    ALLOW_OTHER = 4
    ALLOW_ALL = 6
```

What do you think?

gmazoyer committed:
Rename enum members

gmazoyer committed:
Add permission report to permission backend

gmazoyer committed:
Report attribute permission as int

gmazoyer committed:
Show permission decision value as string

gmazoyer committed:
Workaround enum issue affecting tests

For some reasons the issues on appear when we run tests in a row:
* backend/tests/unit/graphql/test_mutation_create.py::test_create_object_with_node_property
* backend/tests/unit/graphql/test_mutation_create.py::test_create_object_with_single_relationship_node_property
* backend/tests/unit/graphql/test_mutation_create.py::test_create_object_with_multiple_relationships_with_node_property

We should find a better fix for this.

gmazoyer committed:
Replace static strings with dataclass instances

ogenstad commented on backend/infrahub/api/schema.py:
```diff
@@ -247,7 +248,12 @@ async def load_schema(
+                id="",
+                name="",
+                action=GlobalPermissions.MANAGE_SCHEMA.value,
+                decision=PermissionDecision.ALLOW_ALL.value,
```
In the end I think we'd want to check the `branch` variable to see we need the ALLOW_DEFAULT/ALLOW_ALL or ALLOW_OTHER/ALLOW_ALL permission.

ogenstad commented on backend/infrahub/permissions/report.py:
```diff
@@ -2,67 +2,75 @@
+        # What do we do if edit default branch global permission is set?
+        # if not restrict_changes:
+        #     for key in permission_report:
+        #         permission_report[key] |= PermissionDecisionFlag.ALLOW_DEFAULT
```
Can it be an argument to `get_permission_report()`?

ogenstad reviewed (approved):
LGTM. Goes the ALLOW_ALL and ALLOW_OTHER work interchangeably within a non default branch now? I.e. if a user has either of those permissions an action would be allowed?

gmazoyer committed:
Adapt required permission for schema for branch

gmazoyer committed:
Add edit default branch perm to get perm report func

gmazoyer committed:
Disable unused arg (for now)
